stock_by_warehouse/models/product_product.py

```diff
@@ -43,38 +43,20 @@ def _prepare_domain_available_not_reserved(self):
         return domain_quant
 
     @api.multi
-    def _compute_product_available_not_res_dict(self):
-
-        res = {}
-
+    def _compute_qty_available_not_reserved(self):
+        """Method to set the quantity available that is not reserved of a serie of products.
+        """
         domain_quant = self._prepare_domain_available_not_reserved()
-        quants = self.env['stock.quant'].with_context(lang=False).search(
-            domain_quant,
-        ).filtered(lambda x: x.reserved_quantity < x.quantity)
-        # TODO: this should probably be refactored performance-wise
-        for prod in self:
-            vals = {}
-            prod_quant_list = dict(
-                quants.filtered(lambda x: x.product_id == prod).mapped(
-                    lambda x: (x.location_id, x)))
-            quantity = 0
-            for prod_quant in prod_quant_list:
-                quantity += self.env['stock.quant']._get_available_quantity(
-                    prod, prod_quant)
+        quants = self.env['stock.quant'].with_context(lang=False).read_group(
+            domain_quant, fields=['product_id', 'quantity', 'reserved_quantity'], groupby="product_id")
 
-            vals['qty_available_not_res'] = quantity
-            res[prod.id] = vals
-        self._product_available_not_res_hook(quants)
-
-        return res
+        res = {}
+        for quant in quants:
+            quantity = quant.get('quantity') - quant.get('reserved_quantity')
+            res[quant.get('product_id')[0]] = quantity
```

Comment:
Could you add the quantity only if it's positive

Reply:
`res[quant.get('product_id')[0]] = quantity > 0 and quantity or 0.0`

or

```
if quantity > 0:
    res[quant.get('product_id')[0]] = quantity
```

Reply:
The first one but shorter:

`res[quant.get('product_id')[0]] = max(quantity, 0.0)`

Reply:
Done :smile: